stdlib/LinearAlgebra/src/diagonal.jl

```diff
@@ -552,10 +552,9 @@ end
 *(x::Adjoint{<:Any,<:AbstractVector}, D::Diagonal) = Adjoint(map((t,s) -> t'*s, D.diag, parent(x)))
 *(x::Adjoint{<:Any,<:AbstractVector}, D::Diagonal, y::AbstractVector) =
     mapreduce(t -> t[1]*t[2]*t[3], +, zip(x, D.diag, y))
-*(x::Transpose{<:Any,<:AbstractVector}, D::Diagonal) = Transpose(map(*, D.diag, parent(x)))
+*(x::Transpose{<:Any,<:AbstractVector}, D::Diagonal) = Transpose(map((t,s) -> transpose(t)*s, D.diag, parent(x)))
```

Comment:
Does this change the behavior? Was this previously non-recursive?

Reply:
For `Number` eltypes this does not change behaviour, since `transpose` doesn't do anything to them. For block matrices, this is a bugfix. It makes this `Transpose` line consistent with the `Adjoint` line 552 above. Both these are now also covered by tests, testing `Complex` numbers (where the adjoint boils down to conjugation on the element level) and 2x2 matrices (where both `adjoint` and `transpose` have an effect).

Reply:
Here's why I call this a bugfix and think it's worthy of a backport:

```julia
julia> A = reshape([[1 2; 3 4], zeros(Int,2,2), zeros(Int, 2, 2), [5 6; 7 8]], 2, 2)
2×2 Array{Array{Int64,2},2}:
 [1 2; 3 4]  [0 0; 0 0]
 [0 0; 0 0]  [5 6; 7 8]

julia> adjoint(1:2) * A
1×2 Adjoint{Adjoint{Int64,Array{Int64,2}},Array{Array{Int64,2},1}}:
 [1 2; 3 4]  [10 12; 14 16]

julia> transpose(1:2) * A
1×2 Transpose{Transpose{Int64,Array{Int64,2}},Array{Array{Int64,2},1}}:
 [1 2; 3 4]  [10 12; 14 16]

julia> adjoint(1:2) * Diagonal(A)
1×2 Adjoint{Adjoint{Int64,Array{Int64,2}},Array{Array{Int64,2},1}}:
 [1 2; 3 4]  [10 12; 14 16]

julia> transpose(1:2) * Diagonal(A)
1×2 Transpose{Transpose{Int64,Array{Int64,2}},Array{Array{Int64,2},1}}:
 [1 3; 2 4]  [10 14; 12 16]
```